File: src/test/java/store/wckd/server/auth/AuthenticationTests.java
package store.wckd.server.auth;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import reactor.core.publisher.Mono;
import store.wckd.server.entity.User;
import store.wckd.server.factory.Factory;
import store.wckd.server.factory.UserFactory;
import store.wckd.server.repository.UserRepository;
import store.wckd.server.service.UserService;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
public class AuthenticationTests {

    @MockBean
    @Autowired
    private PasswordEncoder passwordEncoder;

    @MockBean
    @Autowired
    private UserService userService;

    private final Factory<User> userFactory;
    private final MockMvc mockMvc;

    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static final String TESTING_ENDPOINT = "/me";

    @Autowired
    public AuthenticationTests(MockMvc mockMvc, UserRepository userRepository) {
        this.mockMvc = mockMvc;

        userFactory = new UserFactory(userRepository);
    }

    @Test
    @DisplayName("It should login properly and return a valid jwt token to be used in the headers")
    public void testLogin() throws Exception {
        when(passwordEncoder.matches(anyString(), anyString()))
                .thenReturn(true);

        Mono<User> userMono = userFactory.createOne();

        when(userService.findByUsername(anyString()))
                .thenReturn(userMono);

        User user = userMono.block();

        assertNotNull(user);

        MockHttpServletRequestBuilder request =
                post(String.format("/login?username=%s&password=%s", user.getUsername(), user.getPassword()))
                        .contentType(MediaType.APPLICATION_JSON)
                        .accept(MediaType.APPLICATION_JSON);

        mockMvc.perform(request)
                .andExpect(status().is(200));
    }
}

File: src/test/java/store/wckd/server/auth/AuthenticationTests.kt
package store.wckd.server.auth

import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.Mockito.anyString
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.post
import store.wckd.server.entity.User
import store.wckd.server.factory.Factory
import store.wckd.server.factory.UserFactory
import store.wckd.server.repository.UserRepository
import store.wckd.server.service.UserService
import org.mockito.Mockito.`when` as every

@SpringBootTest
@AutoConfigureMockMvc
class AuthenticationTests @Autowired constructor(
        private val mockMvc: MockMvc,

        userRepository: UserRepository
) {
    @MockBean
    @Autowired
    private lateinit var passwordEncoder: PasswordEncoder

    @MockBean
    @Autowired
    private lateinit var userService: UserService

    private val userFactory: Factory<User> = UserFactory(userRepository)

    @Test
    @DisplayName("It should login properly and return a valid jwt token to be used in the headers")
    @Throws(Exception::class)
    fun testLogin() {
        every(passwordEncoder.matches(anyString(), anyString()))
                .thenReturn(true)

        val userMono = userFactory.createOne()

        every(userService.findByUsername(anyString()))
                .thenReturn(userMono)

        val user = userMono.block()!!

        mockMvc.post("/login?username=${user.username}&password=${user.password}") {
            contentType = MediaType.APPLICATION_JSON
            accept = MediaType.APPLICATION_JSON
        }.andExpect {
            status { isOk }
        }
    }

    companion object {
        private val objectMapper = ObjectMapper()
        private const val TESTING_ENDPOINT = "/me"
    }
}